wgsl/index.bs

```diff
@@ -6981,13 +6979,296 @@ See [[#statements]] and [[#function-calls]].
+        - A specific variable must always hold a uniform value
+
+        An edge can be understood as an implication from the statement corresponding to its source node to the statement corresponding to its target node.
+
+        To express that something must always be uniform (e.g. the control flow at the call site of a derivative), we simply add an edge from MustBeUniform to the corresponding node.
+        One way to understand this, is that MustBeUniform corresponds to the proposition True, so that MustBeUniform -> X is the same as saying that X is true.
+
+        Reciprocally, to express that we cannot ensure the uniformity of something (e.g. a variable which holds the thread id), we simply add an edge from the corresponding node to CannotBeUniform.
+        One way to understand this, is that CannotBeUniform corresponds to the proposition False, so that X -> CannotBeUniform is the same as saying that X is false.
+
+        A consequence of this interpretation is that every node reachable from MustBeUniform corresponds to something which must be uniform for the program to be valid, and every node from which cannotBeUniform is reachable corresponds to something whose uniformity we cannot guarantee. It follows that we have a uniformity violation (and thus reject the program) if there is any path from MustBeUniform to CannotBeUniform.
+</div>
+
+In more detail, here is what we compute for each function:
+
+* A tag about the control-flow in which the function can be called: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter}
+* A tag for each parameter of the function: one of {AlwaysMustBeUniform, MustBeUniformForControlFlowAfter, MustBeUniformForReturnValue, NoRestriction}
+* A tag for the function as a whole: one of {ControlFlowCannotBeUniformAfter, ReturnValueCannotBeUniform, NoRestriction}
+
+Here is the algorithm for computing this data for a given function:
+
+* Create nodes called "MustBeUniform", "CannotBeUniform", "CF_start", "CF_return", and if the function is non-void a node called "Value_return"
+* Create one node for each parameter of the function which we'll call "arg_i"
+* Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next section.
+* Look at which nodes are reachable from "MustBeUniform"
+    * If this set includes the node "CannotBeUniform", then reject the program.
+    * If this set includes "CF_start", then the control-flow in which the function can be called is AlwaysMustBeUniform.
+    * Otherwise it is MustBeUniformForControlFlowAfter
+    * For each "arg_i" in this set, the corresponding parameter is AlwaysMustBeUniform
+    * Remove from the graph all nodes that have been visited
+* Look at which nodes are reachable from "CF_return"
+    * If this set includes "CannotBeUniform", then the function is ControlFlowCannotBeUniformAfter
+    * For each "arg_i" in this set, the corresponding parameter is MustBeUniformForControlFlowAfter
```

Comment:
Should this be:
```suggestion
    * If this set includes "CannotBeUniform", then the function is ControlFlowCannotBeUniformAfter
    * Otherwise, for each "arg_i" in this set, the corresponding parameter is MustBeUniformForControlFlowAfter
```

If the control flow cannot be uniform after this function what use is there in marking the parameter to claim they must be uniform to regain uniform control flow?

Reply:
By my understanding, the clause on line 7054, asserting a conclusion about the flow uniformity effect occuring *after* the function has been called.
But the clause on 7055 contributes to the uniformity requirement on the value of the i'th argument, to satisfy uniformity demands of builtins executed **during** the lifetime of the function.  So it's not really an "otherwise".

As @Kangz points out, the algorithm as presented here confusingly stitches together the computation of the three kinds of tags.   It gets much more clear if you write out the computation of each tag independently.  I'm currently redrafting in those kinds of terms so I can analyze this better for myself.